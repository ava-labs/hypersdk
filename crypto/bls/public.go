// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

package bls

import (
	"errors"

	"github.com/ava-labs/avalanchego/utils/crypto/bls"
)

const PublicKeyLen = bls.PublicKeyLen

var (
	ErrNoPublicKeys              = errors.New("no public keys")
	ErrFailedPublicKeyDecompress = errors.New("couldn't decompress public key")
)

type (
	PublicKey          = bls.PublicKey
	AggregatePublicKey = bls.AggregatePublicKey
)

func PublicKeyToBytes(pk *PublicKey) []byte {
	return bls.PublicKeyToBytes(pk)
}

func PublicKeyFromBytes(pkBytes []byte) (*PublicKey, error) {
	return bls.PublicKeyFromBytes(pkBytes)
}

func Verify(msg []byte, pk *PublicKey, sig *Signature) bool {
	return bls.Verify(pk, sig, msg)
}
