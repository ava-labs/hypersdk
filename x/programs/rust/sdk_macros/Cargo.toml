[package]
name = "sdk_macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0.67"
quote = "1.0.33"
syn = { version = "2.0.37", features = ["full"] }
unzip-n = "0.1.2"