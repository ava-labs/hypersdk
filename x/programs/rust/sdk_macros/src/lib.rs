extern crate proc_macro;

use core::panic;

use proc_macro::TokenStream;
use proc_macro2::Span;
use quote::{quote, ToTokens};
use syn::{
    parse_macro_input, parse_str, Fields, FnArg, Ident, ItemEnum, ItemFn, Pat, PatType, Type,
};
use unzip_n::unzip_n;
unzip_n!(3);

enum ParamKind {
    SupportedPrimitive,
    Program,
    Pointer,
}

impl From<&Box<Type>> for ParamKind {
    fn from(ty: &Box<Type>) -> Self {
        if is_supported_primitive(ty) {
            ParamKind::SupportedPrimitive
        } else if is_context(ty) {
            ParamKind::Program
        } else {
            ParamKind::Pointer
        }
    }
}

impl ParamKind {
    fn converted_param_tokenstream(&self, param_name: &Ident) -> proc_macro2::TokenStream {
        match self {
            // return the original parameter if it is a supported primitive type
            ParamKind::SupportedPrimitive => {
                quote! {
                    #param_name
                }
            }
            // use the From<i64> trait to convert from i64 to a Program struct
            ParamKind::Program => {
                quote! {
                    #param_name.into()
                }
            }
            // only convert from_raw_ptr if not a supported primitive type or Program
            ParamKind::Pointer => {
                quote! {
                    unsafe { wasmlanche_sdk::state::from_raw_ptr(#param_name) }
                }
            }
        }
    }
}

/// An attribute procedural macro that makes a function visible to the VM host.
/// It does so by wrapping the `item` tokenstream in a new function that can be called by the host.
/// The wrapper function will have the same name as the original function, but with "_guest" appended to it.
/// The wrapper functions parameters will be converted to WASM supported types. When called, the wrapper function
/// calls the original function by converting the parameters back to their intended types using .into().
#[proc_macro_attribute]
pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as ItemFn);
    let name = &input.sig.ident;
    let input_args = &input.sig.inputs;
    let new_name = Ident::new(&format!("{}_guest", name), name.span()); // Create a new name for the generated function(name that will be called by the host)
    let empty_param = Ident::new("ctx", Span::call_site()); // Create an empty parameter for the generated function
    let full_params = input_args.iter().enumerate().map(|(index, fn_arg)| {
        // A typed argument is a parameter. An untyped (receiver) argument is a `self` parameter.
        if let FnArg::Typed(PatType { pat, ty, .. }) = fn_arg {
            // ensure first parameter is Context
            if index == 0 && !is_context(ty) {
                panic!("First parameter must be Context.");
            }

            if let Pat::Ident(ref pat_ident) = **pat {
                let param_name = &pat_ident.ident;
                let param_descriptor = ty.into();
                let param_type = if is_supported_primitive(ty) {
                    ty.to_token_stream()
                } else {
                    parse_str::<Type>("i64")
                        .expect("valid i64 type")
                        .to_token_stream()
                };
                return (param_name, param_type, param_descriptor);
            }
            // add unused variable
            if let Pat::Wild(_) = **pat {
                if is_context(ty) {
                    return (
                        &empty_param,
                        parse_str::<Type>("i64")
                            .expect("valid i64 type")
                            .to_token_stream(),
                        ParamKind::Program,
                    );
                } else {
                    panic!("Unused variables only supported for Program.")
                }
            }
        }
        panic!("Unsupported function parameter format.");
    });

    let (param_names, param_types, converted_params) = full_params
        .map(|(param_name, param_type, param_kind)| {
            (
                param_name,
                param_type,
                param_kind.converted_param_tokenstream(param_name),
            )
        })
        .unzip_n_vec();

    // Extract the original function's return type. This must be a WASM supported type.
    let return_type = &input.sig.output;
    let output = quote! {
        // Need to include the original function in the output, so contract can call itself
        #input
        #[no_mangle]
        pub extern "C" fn #new_name(#(#param_names: #param_types), *) #return_type {
            #name(#(#converted_params),*) // pass in the converted parameters
        }
    };

    TokenStream::from(output)
}

/// This macro assists in defining the schema for a program's state.  A user can
/// simply define an enum with the desired state keys and the macro will
/// generate the necessary code to convert the enum to a byte vector.
/// The enum will automatically derive the Copy and Clone traits. As well as the
/// repr(u8) attribute.
///
/// Note: The enum variants with named fields are not supported.
#[proc_macro_attribute]
pub fn state_keys(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut item_enum = parse_macro_input!(item as ItemEnum);
    // add default attributes
    item_enum.attrs.push(syn::parse_quote! {
         #[derive(Clone, Copy, Debug)]
    });
    item_enum.attrs.push(syn::parse_quote! {
         #[repr(u8)]
    });

    let name = &item_enum.ident;
    let variants = &item_enum.variants;

    let to_vec_tokens = generate_to_vec(variants);
    let gen = quote! {
        // generate the original enum definition with attributes
        #item_enum

        // generate the to_vec implementation
        impl #name {
            pub fn to_vec(self) -> Vec<u8> {
                match self {
                    #(#to_vec_tokens),*
                }
            }
        }
    };

    TokenStream::from(gen)
}

fn generate_to_vec(
    variants: &syn::punctuated::Punctuated<syn::Variant, syn::Token![,]>,
) -> Vec<proc_macro2::TokenStream> {
    variants
        .iter()
        .enumerate()
        .map(|(idx, variant)| {
            let variant_ident = &variant.ident;
            let index = idx as u8;
            match &variant.fields {
                // ex: Point(f64, f64)
                Fields::Unnamed(_) => quote! {
                    Self::#variant_ident(a) => std::iter::once(#index).chain(a.into_iter()).collect()
                },
                // ex: Point
                Fields::Unit => quote! {
                    Self::#variant_ident => vec![#index]
                },
                // ex: Point { x: f64, y: f64 }
                Fields::Named(_) => quote! {
                    Self::#variant_ident { .. } => panic!("named enum fields are not supported"),
                },
            }
        })
        .collect()
}

/// Returns whether the type_path represents a supported primitive type.
fn is_supported_primitive(type_path: &std::boxed::Box<Type>) -> bool {
    if let Type::Path(ref type_path) = **type_path {
        let ident = &type_path.path.segments[0].ident;
        let ident_str = ident.to_string();
        matches!(ident_str.as_str(), "i32" | "i64" | "bool")
    } else {
        false
    }
}

/// Returns whether the type_path represents a Context type.
fn is_context(type_path: &std::boxed::Box<Type>) -> bool {
    if let Type::Path(ref type_path) = **type_path {
        let ident = &type_path.path.segments[0].ident;
        let ident_str = ident.to_string();
        ident_str == "Program"
    } else {
        false
    }
}
