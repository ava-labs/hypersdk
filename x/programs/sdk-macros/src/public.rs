// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

use proc_macro2::{Span, TokenStream};
use quote::{format_ident, quote};
use syn::{
    parse::{Parse, ParseStream},
    parse_quote, parse_str,
    punctuated::Punctuated,
    spanned::Spanned,
    Block, Error, FnArg, Generics, Ident, ItemFn, Pat, PatIdent, PatType, PatWild, ReturnType,
    Signature, Token, Type, TypeReference, Visibility,
};

const CONTEXT_TYPE: &str = "&mut wasmlanche::Context";

type CommaSeparated<T> = Punctuated<T, Token![,]>;

pub fn impl_public(public_fn: PublicFn) -> Result<TokenStream, Error> {
    let args_names = public_fn
        .sig
        .other_inputs
        .iter()
        .map(|PatType { pat: name, .. }| quote! {#name});
    let args_names_2 = args_names.clone();

    let name = &public_fn.sig.ident;
    let context_type = type_from_reference(&public_fn.sig.user_defined_context_type);

    let other_inputs = public_fn.sig.other_inputs.iter();

    let external_call = quote! {
        mod private {
            use super::*;
            #[derive(wasmlanche::borsh::BorshDeserialize)]
            #[borsh(crate = "wasmlanche::borsh")]
            struct Args {
                ctx: #context_type,
                #(#other_inputs),*
            }

            #[link(wasm_import_module = "program")]
            extern "C" {
                #[link_name = "set_call_result"]
                fn set_call_result(ptr: *const u8, len: usize);
            }

            #[no_mangle]
            unsafe extern "C-unwind" fn #name(args: wasmlanche::HostPtr) {
                wasmlanche::register_panic();

                let result = {
                    let args: Args = wasmlanche::borsh::from_slice(&args).expect("error fetching serialized args");

                    let Args { mut ctx, #(#args_names),* } = args;

                    let result = super::#name(&mut ctx, #(#args_names_2),*);
                    wasmlanche::borsh::to_vec(&result).expect("error serializing result")
                };

                unsafe { set_call_result(result.as_ptr(), result.len()) };
            }
        }
    };

    let mut binding_fn = public_fn.to_bindings_fn()?;

    let feature_name = "bindings";

    let mut public_fn = public_fn;

    public_fn
        .attrs
        .push(parse_quote! { #[cfg(not(feature = #feature_name))] });
    binding_fn
        .attrs
        .push(parse_quote! { #[cfg(feature = #feature_name)] });

    public_fn
        .block
        .stmts
        .insert(0, syn::parse2(external_call).unwrap());

    let public_fn = ItemFn::from(public_fn);

    let result = quote! {
        #binding_fn
        #public_fn
    };

    Ok(result)
}

pub struct PublicFn {
    attrs: Vec<syn::Attribute>,
    vis: Visibility,
    sig: PublicFnSignature,
    block: Box<Block>,
}

impl Parse for PublicFn {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let input = ItemFn::parse(input)?;

        let ItemFn {
            attrs,
            vis,
            sig,
            block,
        } = input;

        let vis_err = if !matches!(&vis, Visibility::Public(_)) {
            let err = Error::new(
                sig.span(),
                "Functions with the `#[public]` attribute must have `pub` visibility.",
            );

            Some(err)
        } else {
            None
        };

        let mut inputs = sig.inputs.into_iter();
        let paren_span = sig.paren_token.span.join();

        let (user_defined_context_type, context_input) =
            extract_context_arg(&mut inputs, paren_span);

        let context_input = match (vis_err, context_input) {
            (Some(mut vis_err), Err(context_err)) => {
                vis_err.combine(context_err);
                Err(vis_err)
            }
            (Some(vis_err), Ok(_)) => Err(vis_err),
            (None, context_input) => context_input,
        };

        let other_inputs = map_other_inputs(inputs);

        let (context_input, other_inputs) = match (context_input, other_inputs) {
            (Err(mut vis_and_first), Err(rest)) => {
                vis_and_first.combine(rest);
                Err(vis_and_first)
            }
            (Ok(_), Err(e)) | (Err(e), Ok(_)) => Err(e),
            (Ok(context_input), Ok(other_inputs)) => Ok((context_input, other_inputs)),
        }?;

        let fn_token = sig.fn_token;

        let sig = PublicFnSignature {
            fn_token,
            ident: sig.ident,
            generics: sig.generics,
            paren_token: sig.paren_token,
            user_defined_context_type,
            context_input,
            other_inputs,
            output: sig.output,
        };

        Ok(Self {
            attrs,
            vis,
            sig,
            block,
        })
    }
}

impl From<PublicFn> for ItemFn {
    fn from(public_fn: PublicFn) -> Self {
        let PublicFn {
            attrs,
            vis,
            sig,
            block,
        } = public_fn;

        let sig = sig.into();

        Self {
            attrs,
            vis,
            sig,
            block,
        }
    }
}

#[derive(Clone)]
struct PublicFnSignature {
    ident: Ident,
    generics: Generics,
    fn_token: Token![fn],
    paren_token: syn::token::Paren,
    user_defined_context_type: Box<Type>,
    context_input: ContextArg,
    other_inputs: CommaSeparated<PatType>,
    output: ReturnType,
}

impl From<PublicFnSignature> for Signature {
    fn from(value: PublicFnSignature) -> Self {
        let PublicFnSignature {
            fn_token,
            ident,
            generics,
            paren_token,
            user_defined_context_type: _,
            context_input,
            other_inputs,
            output,
        } = value;

        let inputs = std::iter::once(FnArg::from(context_input))
            .chain(other_inputs.into_iter().map(FnArg::Typed))
            .collect();

        Self {
            constness: None,
            asyncness: None,
            unsafety: None,
            abi: None,
            fn_token,
            ident,
            generics,
            paren_token,
            inputs,
            variadic: None,
            output,
        }
    }
}

#[derive(Clone)]
struct ContextArg {
    pat: Box<Pat>,
    ty: Box<Type>,
    colon_token: Token![:],
}

impl From<ContextArg> for FnArg {
    fn from(value: ContextArg) -> Self {
        let attrs = Vec::new();
        let ContextArg {
            pat,
            ty,
            colon_token,
        } = value;

        FnArg::Typed(PatType {
            attrs,
            pat,
            ty,
            colon_token,
        })
    }
}

impl PublicFn {
    fn to_bindings_fn(&self) -> Result<ItemFn, Error> {
        let Self {
            attrs,
            vis,
            sig,
            block: _,
        } = self;

        let args = sig.other_inputs.iter().map(|PatType { pat, .. }| pat);
        let name = sig.ident.to_string();

        let block = Box::new(parse_quote! {{
            let args = wasmlanche::borsh::to_vec(&(#(#args,)*)).expect("error serializing args");
            ctx
                .call_function(#name, &args)
                .expect("calling the external contract failed")
        }});

        let mut ty = sig.user_defined_context_type.clone();

        if let Type::Reference(ty) = &mut *ty {
            ty.mutability = None;
        }

        let context_input = ContextArg {
            pat: parse_quote! { ctx },
            colon_token: self.sig.context_input.colon_token,
            ty,
        };

        let sig = Signature::from(PublicFnSignature {
            context_input,
            ..sig.clone()
        });

        let item_fn = ItemFn {
            attrs: attrs.clone(),
            vis: vis.clone(),
            sig,
            block,
        };

        Ok(item_fn)
    }
}

fn extract_context_arg<Iter>(
    inputs: &mut Iter,
    paren_span: Span,
) -> (Box<Type>, Result<ContextArg, Error>)
where
    Iter: Iterator<Item = FnArg>,
{
    let mut user_defined_context_type: Box<Type> = Box::new(parse_str(CONTEXT_TYPE)).unwrap();

    let first_arg = inputs.next();

    let Some(first_arg) = first_arg else {
        let err = Error::new(
                    paren_span,
                    format!("Functions with the `#[public]` attribute must have at least one parameter and the first parameter must be of type `{CONTEXT_TYPE}`"),
                );

        return (user_defined_context_type, Err(err));
    };

    let result = match first_arg {
        FnArg::Typed(PatType {
            mut ty,
            pat,
            colon_token,
            ..
        }) if is_mutable_context_ref(&ty) => {
            let types = (user_defined_context_type.as_mut(), ty.as_mut());

            if let (Type::Reference(context_type), Type::Reference(ty)) = types {
                context_type.lifetime = ty.lifetime.clone();
            }

            // swap the fully qualified `Context` with the user-defined `Context`
            // for better compiler errors
            std::mem::swap(types.0, types.1);

            Ok(ContextArg {
                pat,
                ty,
                colon_token,
            })
        }

        first_arg => {
            let message = format!("The first paramter of a function with the `#[public]` attribute must be of type `{CONTEXT_TYPE}`");

            let span = match first_arg {
                FnArg::Typed(PatType { ty, .. }) => ty.span(),
                _ => first_arg.span(),
            };

            Err(Error::new(span, message))
        }
    };

    (user_defined_context_type, result)
}

fn map_other_inputs(inputs: impl Iterator<Item = FnArg>) -> Result<CommaSeparated<PatType>, Error> {
    let other_inputs = inputs.enumerate().map(|(i, fn_arg)| match fn_arg {
        FnArg::Receiver(_) => Err(Error::new(
            fn_arg.span(),
            "Functions with the `#[public]` attribute cannot have a `self` parameter.",
        )),
        FnArg::Typed(mut pat_type) => {
            let pat = pat_type.pat;
            pat_type.pat = match *pat {
                // replace wildcards with generated identifiers
                Pat::Wild(PatWild { attrs, .. }) => Pat::Ident(PatIdent {
                    attrs,
                    by_ref: None,
                    mutability: None,
                    ident: format_ident!("arg{}", i),
                    subpat: None,
                }),
                pat => pat,
            }
            .into();

            Ok(pat_type)
        }
    });

    // this isn't the same thing as a `try_fold` because we don't exit early on an error
    #[allow(clippy::manual_try_fold)]
    other_inputs.fold(Ok(Punctuated::new()), |result, param| {
        match (result, param) {
            // ignore Ok or first error encountered
            (Err(errors), Ok(_)) | (Ok(_), Err(errors)) => Err(errors),
            // combine errors
            (Err(mut errors), Err(e)) => {
                errors.combine(e);
                Err(errors)
            }
            // collect results
            (Ok(mut names), Ok(name)) => {
                names.push(name);
                Ok(names)
            }
        }
    })
}

/// Returns whether the type_path represents a mutable context ref type.
fn is_mutable_context_ref(type_path: &Type) -> bool {
    let Type::Reference(TypeReference {
        mutability: Some(mutability),
        elem,
        ..
    }) = type_path
    else {
        return false;
    };

    // span is ignored in the comparison
    if mutability != &Token![mut](mutability.span()) {
        return false;
    }

    if let Type::Path(type_path) = elem.as_ref() {
        let context_path = parse_str::<TypeReference>(CONTEXT_TYPE).unwrap();
        let Type::Path(context_path) = context_path.elem.as_ref() else {
            return false;
        };

        let context_ident = context_path
            .path
            .segments
            .last()
            .map(|segment| &segment.ident);
        type_path.path.segments.last().map(|segment| &segment.ident) == context_ident
    } else {
        false
    }
}

fn type_from_reference(type_path: &Type) -> &Type {
    if let Type::Reference(TypeReference { elem, .. }) = type_path {
        elem.as_ref()
    } else {
        type_path
    }
}
