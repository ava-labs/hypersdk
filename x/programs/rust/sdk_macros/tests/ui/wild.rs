use sdk_macros::public;

struct Program;

#[public]
pub fn test(_: Program, _: usize) {}

fn main() {}
