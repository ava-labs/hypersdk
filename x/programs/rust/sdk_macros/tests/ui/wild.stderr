error: Functions with the `#[public]` attribute can only ignore the first parameter.
 --> tests/ui/wild.rs:6:25
  |
6 | pub fn test(_: Program, _: usize) {}
  |                         ^
